### PR TITLE
Fix of a view of empty trend chart when there is no data for it

### PR DESCRIPTION
**Pull request by Vitalii-Misechko**
Fix of a view of empty trend chart when there is no data for it

* Functional Testing 
  * An empty trend chart has ticks for y-axis even there is no data
  * A trend chart shows correct ticks when trend data is provided
  * Filters correctly hides/shows trend data
  
* Security, devops, perf: N/A
* UX and docs
  * [x] Demo
  
FYI @anhill-D2L @hyehayes @johngwilkinson @NicholasHallman @bemailloux @MykolaGalian 

### Before

image

### After

image

### Demo

Ticks-Demo-For-Empty-NonEmpty-chart

### CODE REVIEW

File: components/access-trend-card.js [hunk 1]
```diff
@@ -272,8 +289,18 @@ class AccessTrendCard extends SkeletonMixin(Localizer(MobxLitElement)) {
...
+
+	_onChartLoad(event) {
+		this._chart = event.detail;
+	}
+
 	get _series() {
-		if (!this.data._data) return [];
+		if (this._emptyData) return [0, 0];
+
+		if (this._chart) this._chart.redraw(); // update tickPositions for the chart with data
```
Comment (johngwilkinson) [inline components/access-trend-card.js:302]: Why is this necessary, and could it go in `render` instead of in a getter? Seems dangerous to rely on a side-effect here.

**Review by bemailloux — COMMENTED** [review]
What happens if you load the chart with data initially, then change the filters such that the chart has no data? Does it show the empty grid in that case?

File: components/access-trend-card.js (hunk 1 above)
Comment (Vitalii-Misechko) [reply components/access-trend-card.js:303]: Good point!  Removed, it works without calling `redraw`

**Comment by Vitalii-Misechko** [comment]
> What happens if you load the chart with data initially, then change the filters such that the chart has no data? Does it show the empty grid in that case?

Unfortunately, it doesn't. It shows grid with a previous series. It seem one more bug (

image

Looking into it

**Comment by Vitalii-Misechko** [comment]
> > What happens if you load the chart with data initially, then change the filters such that the chart has no data? Does it show the empty grid in that case?
> 
> Unfortunately, it doesn't. It shows grid with a previous series. It seem one more bug (
> 
> image
> 
> Looking into it

Fixed

File: components/access-trend-card.js [hunk 2]
```diff
@@ -272,8 +289,16 @@ class AccessTrendCard extends SkeletonMixin(Localizer(MobxLitElement)) {
 		return orgUnit ? orgUnit[ORG_UNIT.NAME] : '';
 	}
 
+	get _emptyData() {
+		return !this.data._data || this._trendData.length === 0;
+	}
+
+	_onChartLoad(event) {
+		this._chart = event.detail;
```
Comment (johngwilkinson) [inline components/access-trend-card.js:297]: is `this._chart` still used anywhere?

**Review by bemailloux — APPROVED** [review]

File: components/content-views-card.js [hunk 3]
```diff
@@ -248,8 +254,16 @@ class ContentViewsCard extends SkeletonMixin(Localizer(MobxLitElement)) {
 		return orgUnit ? orgUnit[ORG_UNIT.NAME] : '';
 	}
 
+	get _emptyData() {
+		return !this.data._data || !this.userData.contentViews || this._trendData.length === 0;
+	}
+
+	_onChartLoad(event) {
+		this._chart = event.detail;
```
Comment (johngwilkinson) [inline components/content-views-card.js:262]: used?

**Review by johngwilkinson — COMMENTED** [review]
At least one of the new goldens in this PR is bad (empty cards). I also notice the y-axis ticks look very close together in the new ones - is that intentional?

File: components/access-trend-card.js (hunk 2 above)
Comment (Vitalii-Misechko) [reply components/access-trend-card.js:297]: nope, can be deleted

**Comment by Vitalii-Misechko** [comment]
> At least one of the new goldens in this PR is bad (empty cards). 
I'll take a look. I considered this empty card as some kind of flakyness.

> I also notice the y-axis ticks look very close together in the new ones - is that intentional?

Removed `intervalTick: 1`, it is supposed to prevent fractional ticks on Course Access and Content View trend charts. But, it's better to discuss all pros and cons of the solution with Kevin first.

File: components/content-views-card.js (hunk 3 above)
Comment (Vitalii-Misechko) [reply components/content-views-card.js]: removed

**Review by johngwilkinson — APPROVED** [review]

**Comment by Vitalii-Misechko** [comment]
> At least one of the new goldens in this PR is bad (empty cards).

I've just gone through golden snapshots and I haven't found a bad card. Either the golden is fixed or I overlooked it.

Could you point me out if the bad snapshot exists?

**Comment by johngwilkinson** [comment]
> > At least one of the new goldens in this PR is bad (empty cards).
> 
> I've just gone through golden snapshots and I haven't found a bad card. Either the golden is fixed or I overlooked it.
> 
> Could you point me out if the bad snapshot exists?

The ones you just merged look fine. They are different than the ones that were here before.

**Comment by Vitalii-Misechko** [comment]
I'll port these changes to the master tomorrow if there are no objections )

**Comment by johngwilkinson** [comment]
In general: for flaky tests, we shouldn't merge the wrong goldens, because then our tests are trying to prove that our code produces those results.

**Comment by Vitalii-Misechko** [comment]
> In general: for flaky tests, we shouldn't merge the wrong goldens, because then our tests are trying to prove that our code produces those results.

I agree with that. I just didn't figure out what to do with such tests. Re-run or close visual diff PR and re-run it?

**Comment by johngwilkinson** [comment]
> > In general: for flaky tests, we shouldn't merge the wrong goldens, because then our tests are trying to prove that our code produces those results.
> 
> I agree with that. I just didn't figure out what to do with such tests. Re-run or close visual diff PR and re-run it?

Just re-run the github jobs, I think.
